Added convenience methods for passing stencils to functions

ikottlarz opened the pull request:
Hi there!

I added a few convenience methods for passing stencils to `SpatialSymbolicPermutation`, namely:

- for 2D and 3D, pass rectangular/ cuboid stencils just via their `extent` (an extension of the word length in 1d, just how many points should be considered in each direction, so that `word_length = prod(extent)`) and a lag in each direction as `SpatialSymbolicPermutation(extent, lag, x)`, such that
  ```julia
  est = SpatialSymbolicPermutation((2, 2), (2, 2), x)
  ```
  is equivalent to the original
  ```julia
  stencil = CartesianIndex.([(0, 2), (2, 0), (2, 2)])
  est = SpatialSymbolicPermutation(stencil, x)
  ```
- In general, pass n-dimensional `Array`s instead of a list of cartesian indices, where for example
  ```julia
  stencil = [1 0 1;
             0 0 0; 
             1 0 1]
  ```
  would correspond to the cartesian-index version
  ```julia
  stencil = CartesianIndex.([(0, 2), (2, 0), (2, 2)])
  ```

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -21,13 +23,34 @@ stencil = CartesianIndex.([(0,1), (1,1), (1,0)])
 within the stencil dictates the order that pixels are compared with.
 The pixel without any offset is always first in the order.
 
+For rectangular/ cuboid stencils, one can also pass two `NTuple`s `extent` and `lag`,
```
In this sentence, perhaps mention that for these tuples,`N` in `NTuple{N, T}` must match the dimension of the data too. You do mention thus explicitly for the integer array approach, but not for the `extent` and `lag` approach.

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
````diff
@@ -21,13 +23,34 @@ stencil = CartesianIndex.([(0,1), (1,1), (1,0)])
+```julia
+data = [rand(50, 50) for _ in 1:50]
+x = data[1] # first "time slice" of a spatial system evolution
+stencil = [1 1; 1 1]
````
From further down the docstring: "... no offset (meaning the pixel itself) is always included automatically". How is this handled in the case of integer array stencils, and in the case of `extent` and `lag` stencils?

Does it mean, for example for the 2D case, that the stencil `[0 1; 1 1]` would not be possible?

ikottlarz replied:
Your example would be possible, the offset doesn't have to be in the upper left corner of the matrix. The way I generate the stencils from the input array, I make sure to translate them in a way that the "no offset" coordinate can be included. Your example would be translated into `CartesianIndex.([(0,1), (-1,1)])` (plus the `(0,0)` that will later be added).

kahaaga replied:
> Your example would be possible, the offset doesn't have to be in the upper left corner of the matrix. The way I generate the stencils from the input array, I make sure to translate them in a way that the "no offset" coordinate can be included. Your example would be translated into CartesianIndex.([(0,1), (-1,1)]) (plus the (0,0) that will later be added).

Excellent! Perhaps then the docstring should say that if using integer arrays as stencils, then the "pixel itself is included" part doesn't apply - instead, what you demand is what you get - or something along those lines?

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -77,6 +100,43 @@ function SpatialSymbolicPermutation(
     SpatialSymbolicPermutation{D, p, typeof(valid)}(stencil, copy(stencil), arraysize, valid)
 end
 
+function SpatialSymbolicPermutation(
```
If the `extent` and `lag` approach is only implemented for 2D and 3D, this must be mentioned in the docstring. 

Even better: we should perhaps consider just use some generated code here to generalize to higher dimensions without code duplication.

ikottlarz replied:
> Even better: we should perhaps consider just use some generated code here to generalize to higher dimensions without code duplication.

Good point, I had been thinking about that too. I implemented the generalized way now

kahaaga replied:
> Good point, I had been thinking about that too. I implemented the generalized way now

Excellent.

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -21,13 +23,34 @@ stencil = CartesianIndex.([(0,1), (1,1), (1,0)])
 
+For rectangular/ cuboid stencils, one can also pass two `NTuple`s `extent` and `lag`,
+from which an appropriate stencil will be created. 
+`extent` defines how many points should be considered in each direction, and `lag`
```
maybe say "`extent[i]` defines how many points should be considered along the `i`-th axis, and `lag[i]` ..."

kahaaga replied:
The points *are* actually considered, so we should write "... defines how many points *are* considered", not the thinking-about-potentially-doing-so-if-I-decide-to-do-so-in-the-future "...defines how many points *should be* considered" :D

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -21,13 +23,34 @@ stencil = CartesianIndex.([(0,1), (1,1), (1,0)])
+For rectangular/ cuboid stencils, one can also pass two `NTuple`s `extent` and `lag`,
+from which an appropriate stencil will be created. 
+`extent` defines how many points should be considered in each direction, and `lag`
+defines the offset between them.
```
Defines the offset between *what*? The points? Per-dimension? How? I'm not able to decipher this statement. This perhaps needs a short description of precisely what is meant.

kahaaga reviewed (commented):
Thanks for the PR, @ikottlarz! I have mostly documentation-related comments. I suggest you wait for @Datseris' comments too before you make any substantial changes, because we might disagree on some things.

The only thing I'd really like to see is generated code for higher dimensions. But I guess for most use cases 2D and 3D is sufficient, so I'm happy to approve this as is. Someone can submit an improvement PR with generated code whenever somebody has time to do so.

Datseris reviewed (commented):
Thanks for the contribution! 

Alright, this looks good. Here is what I propose as the API:

1. We have a **single method `SSP(stencil, x, periodic = true)`**.
2. In the documentation, via means of an _enumerated list_ we state the **three possibiilities** of `stencil`: Vector of cartesian indices of offsets, array with boolean elements, a TUPLE of TUPLES (extent, lag).
3. **Let's remove the "0-offset is assumed".** When I wrote the first version, where cartesian offsets was the only way, it made sense to define stenciles this way. Now, with the two new methods, it unecessarily complicates things. So, it's better to not impose this "0 offset is imposed". We simply add a comment to the vector of cartesian offsets methods like "don't forget to include 0 offset if you want to include the point itself, which is almost always the case."

kahaaga commented:
> Thanks for the contribution!

> Alright, this looks good. Here is what I propose as the API:
> We have a single method SSP(stencil, x, periodic = true).
> In the documentation, via means of an enumerated list we state the three possibiilities of stencil: Vector of cartesian indices of offsets, array with boolean elements, a TUPLE of TUPLES (extent, lag).
> Let's remove the "0-offset is assumed". When I wrote the first version, where cartesian offsets was the only way, it made sense to define stenciles this way. Now, with the two new methods, it unecessarily complicates things. So, it's better to not impose this "0 offset is imposed". We simply add a comment to the vector of cartesian offsets methods like "don't forget to include 0 offset if you want to include the point itself, which is almost always the case."
> These changes should significantly reduce the amount of "discussion" in the documentation string.

This is a good compromise. 

@ikottlarz Whats left to do then is:

- [ ] Let `extent`-`lag`-method to accept a `Tuple{NTuple{N, T}} where {N, T}` instead.
- [ ] Write/update the new docstring, if you're up for it. I will have a last look and address any wording issues for the documentation before merging, but I think working on the documentation over several iterations increases its quality, so feel free to give it a go. Otherwise, either me or @Datseris will do it.

ikottlarz commented:
@Datseris  @kahaaga thank you for your feedback! I have now
- changed the signature of the `extent`/`lag` version
- had a go at the docstring
- removed the auto-addition of the zero-offset (did I understand correctly that this was the way we want to go now?)

I also have a two comments to the part of the docstring for the `CartesianIndex` version. It says there
> The length of the stencil decides the order of the permutation entropy, and the ordering within the stencil dictates the order that pixels are compared with.


1. "order of the entropy" makes it sound a bit like a "q-order entropy", which is not what's meant here. The length of the stencil is the "word length", or symbol length.
2. Please correct me if I'm wrong, but I think the ordering in which the pixels are compared doesn't actually matter for the entropy. The symbols are all permutations of one another, and if you change the order in which the pixels are are compared, you permute the bins of the histogram, but since the summation over the probabilities for the entropy is commutative, this doesn't matter for the value of the PE (or other commonly used statistical measures, like the statistical complexity by Rosso et al.).

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+consider, and compute the ordinal pattern within the stencil.
+Stencils are passed in one of the following three ways:
+
+1. as vectors of `CartesianIndex` which encode the pixels to include in the
```
Write `As`, not `as` for points 1., 2., and 3. These are complete sentences, so should all start with a capitalized letter.

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    The pixel without any offset is always first in the order.
+
+2. as a `D`-dimensional array (where `D` matches the dimensionality of the input data)
+    containing `0`s and `1`s, where every pixel with a one is included.
```
*"... `0`s and `1`s, where every pixel with a one is included"* is a bit vague.

I'd rather be more precise, and say something like *"... `0`s and `1`s, where if `stencil[i1, i2, …, iD] == 1`, then the corresponding pixel is included, and if `stencil[i1, i2, …, iD] == 0`, then it is not included."*

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
-The length of the stencil decides the order of the permutation entropy, and the ordering
-within the stencil dictates the order that pixels are compared with.
-The pixel without any offset is always first in the order.
+consider, and compute the ordinal pattern within the stencil.
```
The latter part of the following sentence is not complete: *"A _stencil_ defines what local area around each pixel to consider, and compute the ordinal pattern within the stencil.*

Perhaps something like the following is better: *"A _stencil_ defines what local area (which points) around each pixel to consider and compute ordinal patterns from."*

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
````diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    stencil = CartesianIndex.([(0,0), (0,1), (1,1), (1,0)])
+    est = SpatialSymbolicPermutation(stencil, x)
+    ```
+    Don't forget to include 0 offset if you want to include the point itself, 
````
"Don't forget to include **the zero offset index** if you want to include the point itself, ..." (complete sentence), not "Don't forget to include 0 offset..." (incomplete sentence).

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+
+2. as a `D`-dimensional array (where `D` matches the dimensionality of the input data)
+    containing `0`s and `1`s, where every pixel with a one is included.
+    To generate the same estimator as in 1., for example
```
"To generate the same estimator as in 1., for example" is an incomplete sentence. I think a word or two went missing here!

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
````diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    est = SpatialSymbolicPermutation(stencil, x)
+    ```
+
+3. as a `Tuple` containing two `Tuple`s of length `D` for `D`-dimensional data.
````
A bit more precise would be to say: *"As a `Tuple` containing two `Tuple`s, **both of length `D`**, for `D`-dimensional data"*

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+3. as a `Tuple` containing two `Tuple`s of length `D` for `D`-dimensional data.
+    The first tuple specifies the `extent` of the stencil, where `extent[i]` 
+    dictates the number of pixels to be included along the `i`th axis and `lag[i]`
+    their respective separation.
```
"their respective separation" is ambiguous - does "respective" refer to the separation of axes or pixels? I know this is obvious for the code authors, but maybe not for fresh users. 

I'd suggest *"..., `extent[i]` dictates the number of pixels to be included along the `i`th axis, and `lag[i]` dictates the separation of pixels along the same axis."*

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -10,23 +10,50 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    The first tuple specifies the `extent` of the stencil, where `extent[i]` 
+    dictates the number of pixels to be included along the `i`th axis and `lag[i]`
+    their respective separation.
+    This method can only generate cuboid stencils. To create the same estimator as
```
I don't think [cuboid](https://en.wikipedia.org/wiki/Cuboid) is the word to use here. Maybe say *"This method can only generate (hyper)rectangular stencils"* instead?

ikottlarz replied:
Yes, that was the word I was looking for I think :)

kahaaga commented on test/estimators/permutation_spatial.jl:
```diff
@@ -46,4 +59,21 @@ using Entropies, Test
     ps = probabilities(w, est)
     @test length(ps) > 1
+
+    # check that the 3d-cuboid version also works as expected
```
Again, "rectangle"/"hyperrectangle" or "square"/"hypersquare", not cuboid.

kahaaga commented:
Thanks for the fixes, @ikottlarz! I added some more comments on the language, which it would be nice if you could fix.

> "order of the entropy" makes it sound a bit like a "q-order entropy", which is not what's meant here. The length of the stencil is the "word length", or symbol length.

I agree that the wording is a bit ambiguous. Since we now have three different stencil types, "length of the stencil" in itself is ambiguous, because in the case of providing cartesian coordinates it would be true, but when providing a stencil like `[0 1; 1 1]`, the length of the array does not longer correspond to the number of points that are actually included within the stencil.

 Would it be more precise to, instead of  `"The length of the stencil decides the order of the permutation entropy"`, say something like

*"The number included points in a stencil (`m`) determines the length of the vectors to be symbolized, i.e. there are `m!` possible ordinal/permutation patterns around each pixel"*? Perhaps a bit long, but it precisely says what's going on. I'm open for other suggestions.

One could also perhaps refer to `m` in for example the `SymbolicPermutation` estimator, but I don't know if that causes more confusion than it clarifies things. I'm not sure.

ikottlarz commented:
Thank you for the comments @kahaaga! I mostly went with your suggestions in the updated docstring. Is there anything else from your side? Does @Datseris have any requests for changes in these alterations?

kahaaga commented:
> Please correct me if I'm wrong, but I think the ordering in which the pixels are compared doesn't actually matter for the entropy. The symbols are all permutations of one another, and if you change the order in which the pixels are are compared, you permute the bins of the histogram, but since the summation over the probabilities for the entropy is commutative, this doesn't matter for the value of the PE (or other commonly used statistical measures, like the statistical complexity by Rosso et al.).

Correct me if I'm wrong too :D, but I think the ordering *does* matter, but *only within each stencil* (that is, in the local area around each point). The exception would be for uniformly distributed noise, for which, on average, the choice of stencil doesn't matter (at least for stencils yielding the same-dimensional vectors for symbolization).

Take for example the image:

```julia
3×3 Matrix{Float64}:
 0.6  0.7  0.1
 0.7  1.0  0.9
 1.1  0.8  0.3
```

The stencil `[0 1; 1 1]` would yield vectors `(0.7, 0.7, 1.0), (1.1, 1.0, 0.8), (1.0, 0.1, 0.9), (0.8, 0.9, 0.3)`. With `encode_motif`, the symbols for this image is `[0, 5, 4, 3]`.
The stencil `[1 1; 1 0]` would yield vectors `(0.6, 0.7, 0.7), (0.7, 1.1, 1.0), (0.7, 1.0, 0.1), (1.0, 0.8, 0.9)`.With `encode_motif`, the symbols for this image is `[0, 1, 3, 4]`.

With `Entropies.encode_motif`, the resulting symbols for this example image is, 
- with stencil `[0 1; 1 1]`: symbols are `[0, 5, 4, 3]`
- with stencil `[1 1; 1 0]`: symbols are `[0, 1, 3, 4]`

So, depending on the order of the points within the stencil, *different* symbols are obtained for *the same pixel*.

However, if the statement refers to the step *after* symbolization, the order in which the symbols (one per pixel) are compared *does not* matter, because the probability `p[i]` is just the normalized frequency of the `i`th symbol (which doesn't care about ordering).




The corresponding

ikottlarz commented:
@kahaaga Ah, I worded my point poorly. I agree that _this_ ordering matters. What does not matter (and what I understood was meant by that part of the docstring, because after that there was also a sentence about the zero-offset one always being the first in that order) was in which order the pixels appear in the vector of `CartesianIndex`. Like,
```julia
stencil = CartesianIndex.([(0,0), (0,1), (-1,1)])
```
would yield the same PE as 
```julia
stencil = CartesianIndex.([(0,0), (-1,1), (0,1)])
```
This is something that is not captured anyway by the two new approaches to passing a stencil anyway. But maybe I just didn't understand correctly what was meant here.

kahaaga commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -9,24 +9,52 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    which is almost always the case.
+    Here the stencil creates a 2x2 square extending to the bottom and right of the pixel
+    (directions here correspond to the way Julia prints matrices by default).
+    The number of included points in a stencil (`m`) determines the length of the vectors
```
I think maybe this sentence *"The number of included points in a stencil (`m`) determines the length of the vectors to be symbolized, i.e. there are `m!` possible ordinal patterns around each pixel."* belongs *before* the enumerated list, because this is a statement that is true of all three stencil types. Then we should say how `m` is computed for each stencil type under the respective enumerated point.

For method 1, `m = length(stencil)`. For method 2 `m = count(!iszero, stencil)`. Is it possible to say what `m` would be a priori for method 3 too, @ikottlarz ?

ikottlarz replied:
Yes, for the 3rd method it's just `m = prod(extent)!`. I'll add it to the docstring!

kahaaga replied:
@ikottlarz Does `prod(extent)!` also take into consideration the `lags`?

kahaaga reviewed (changes requested):
I'd like for you to consider one last change, @ikottlarz, and that regards the computation of the vector length `m` for method 3 and listing `m` for the three different methods.

kahaaga commented:
> @kahaaga What does not matter [...] was in which order the pixels appear in the vector of CartesianIndex

Hm, now I'm slightly confused again. 

Do we agree @ikottlarz, based on my example above, that

```julia
stencil = CartesianIndex.([(0,0), (0,1), (-1,1)])
```
is in fact *not* equivalent to the following?

```julia
stencil = CartesianIndex.([(0,0), (0,1), (-1,1)])
```

ikottlarz commented:
@kahaaga your examples are exactly the same, I think you meant to change something in one of them?

Your two examples above would be something like
```julia
stencil = CartesianIndex.([(0,0), (0,1), (-1,1)])
```
(for the first one) and
```julia
stencil = CartesianIndex.([(0,0), (1,0), (-1,1)])
```
(for the second one). They are not equivalent, there's one cartesian index that only appears in the first (`(0,1)`) and one that appears only in the second (`(1,0)`). But it doesn't matter in which order I insert the cartesian indices into the vector, that's what I was talking about.

kahaaga commented:
@ikottlarz, Sorry, you're right, I messed up the example. What I meant to ask about was the ordering of cartesian indices in the stencil.

In the following example, the cartesian indices for both `a` and `b` are the same.

```julia
a = CartesianIndex.([(0,0), (0,1), (-1,1)])
b = CartesianIndex.([(0,0), (-1,1), (0,1)])
```

However, the stencil `a` would *not* be equivalent to stencil `b`, because the order in which points are picked are different. If around some fixed pixel, stencil `a` would yield the vector `[1, 5, -3]` (whose permutation pattern is `3, 1, 2`), then stencil `b` for the same pixel would yield the vector `[1, -3, 5]` (whose permutation pattern is `2, 1, 3`). Hence, they would yield different symbols for the same pixel. That means that `a` and `b` are not equivalent, and that the ordering in which one inserts the cartesian indices into the stencil must matter.

Am I understanding this correctly?

kahaaga commented:
> @kahaaga yes, the permutation patterns change.

Ok, then we're on the same page! As long as the *code* does result in different permutation patterns (and thus symbols) for differently-ordered stencils, all is good. 

However that might affect statistical properties of some analysis is more of a meta-discussion not relevant to this PR :D 

> The corresponding histogram has the same statistical properties either way (except things like curtosis and so on, which I've never seen being considered in this context anyway since it's a discrete distribution). So it doesn't matter for the results you want to extract from the data (or do you have an example where it does? I've had a discussion about this some time ago for the 2d case and we came to the conclusion that it doesn't matter, at least not for the properties we considered, which were PE and statistical complexity).

I'll have to think about this a bit more, but from first glance, your argument seems to make sense.

This PR is good to go from my side. Anything you'd like to add, @Datseris ?

Datseris commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
````diff
@@ -9,24 +9,55 @@ The input data `x` are high-dimensional arrays, for example 2D arrays [^Ribeiro2
+    as the data. For example
+
+    ```julia
+    data = [rand(50, 50) for _ in 1:50]
+    x = data[1] # first "time slice" of a spatial system evolution
+    stencil = CartesianIndex.([(0,0), (0,1), (1,1), (1,0)])
+    est = SpatialSymbolicPermutation(stencil, x)
````
The lines `data, x, est` don't need to be duplicated in each of the three points. You can put them before the numbered list, and use the numbered list to specify the stencil itself.

Datseris commented on src/probabilities_estimators/permutation_ordinal/spatial_permutation.jl:
```diff
@@ -77,6 +106,29 @@ function SpatialSymbolicPermutation(
+    SpatialSymbolicPermutation(stencil, x, p)
+end
+
+function SpatialSymbolicPermutation(
```
Instead of repeating `SpatialSymbolicPermutation` three times, it makes for simpler design, and more future proof, to define an internal function `stencil_to_offsets` that dispatches on stencil type and always gives the vector of cartesians. in the case where stencil is already vector cartesisans the functio nreturns its argument.

kahaaga replied:
Agreed.

Datseris reviewed (approved):
Very good! Two simple comments only!

ikottlarz commented:
@Datseris thanks for the suggestions, I included them!
